src/vanna/__init__.py

```diff
@@ -184,13 +185,15 @@
 )
 from .utils import sanitize_model_name, validate_config_path
 
+_logger = initialize_logger()
```

Comment:
you should not configure the output of the logger at the global level. Then as soon as someone inports this module, the logger will start printing to stdout. Instead, configure it as one of the first things in the actual main() function.

Also, if this is the only place where `initialize_logger()` is used, it doesn't need to be in that distant module, but just define it directly here.

Reply:
> Then as soon as someone inports this module, the logger will start printing to stdout

I thought the goal was to standardize all prints to the new logging solution?

Reply:
hmm I see what you're saying. I think it just comes down to what we want as the default:

- default to not configuring the logger, and then users opt-in to logging with `logging.basicConfig(...)` or the more finetipped `logging.getLogger("vanna").do_stuff(...)`.
- default to configuring the logger, and have users opt-out.

Currently the behavior is default-on. But arguably the reason for that is just because we can't turn print() off. Some libraries go for default on (eg [splink](https://github.com/moj-analytical-services/splink)), others go for default off, [flask is more nuanced](https://flask.palletsprojects.com/en/2.3.x/logging/#default-configuration). I prefer quietness, so I prefer to not do anything in library code, and let the end app developer configure logging. But I see the pros/cons of all 3.

Reply:
hmm, this is good reading: https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library

Reply:
quote from there: 

Note It is strongly advised that you do not add any handlers other than [NullHandler](https://docs.python.org/3/library/logging.handlers.html#logging.NullHandler) to your library’s loggers. This is because the configuration of handlers is the prerogative of the application developer who uses your library. The application developer knows their target audience and what handlers are most appropriate for their application: if you add handlers ‘under the hood’, you might well interfere with their ability to carry out unit tests and deliver logs which suit their requirements.

So at least I think we should lean on the conservative side of adding handlers, maybe doing something similar to flask?

Reply:
For all CLI/main entrypoints of this lib, we should configure logging of course, so behavior will stay similar to today. But if I'm using vanna as a lib, this is when behavior would change.

Reply:
Hmm, I see what you mean. It was a little unclear to me what the you wanted to solution to look like. But still, we do not want `print()` inside the framework, as we cannot mute these. A good example is that when training Vanna, the documents and SQLs will be printed directly in the console. This is unfortunate for production use.

I'm fine with removing the `_logger` solution given the new insight, but I guess I could revert to just doing `logging.info()` (and similar) and then it is up to the end user to change the logging behaviour.

I can make an attempt at this later today. A bit busy atm.

Reply:
Really appreciate the work here. I am rereading my review and it sounds a bit harsh, I could have been more graceful, so I'm sorry if I came off as a jerk.

Maybe wait for @zainhoda to chime in before you go through the rewrite? They might have a different idea totally for what they want the result to look like.

Reply:
> Really appreciate the work here. I am rereading my review and it sounds a bit harsh, I could have been more graceful, so I'm sorry if I came off as a jerk.

@NickCrews No worries! I figured it was late in the day ;) Have done some open-source and learned to take some criticism on my bad ideas and code suggestions. So its all good!

> Maybe wait for @zainhoda to chime in before you go through the rewrite? They might have a different idea totally for what they want the result to look like. 

Yeah, that sounds like a good idea. I am in no rush. Just ping me if I fail to see an update :]